### ISSUE TITLE
How to specify training time

### ISSUE THREAD

**Issue by ballcap231** [ludwig-ai/ludwig#314]
I'm following the documentation for using the ludwig API but I can't find any guide on how to specify training time. When I run the example code for the ludwig API, training runs for about 1 minute or 2 and then stops. How can I run the training longer?

Here is my simple code:
os.chdir('/content/temp_drive/My Drive/ludwig_code')
model_definition = {}
model_definition['input_features'] = [{'name': i, 'type': 'numerical'} for i in list(df.columns[x] for x in range(4))]
model_definition['output_features'] = [{'name':'OUTPUT1','type':'numerical'}]
model_definition['training'] = {}
ludwig_model = LudwigModel(model_definition)
train_stats = ludwig_model.train(data_train_df=df)
#yaml.dump(model_definition)

**Comment by w4nderlust** [comment]
Hi @ballcap231 . There is no concept of training time. Ludwig trains for certain amount of epochs and performs early stopping if a specific measure of performance on the validation set for a certain amount of epoch. In the [training section of the user guide](https://uber.github.io/ludwig/user_guide/#training) you can find all the parameters and you can change them to your liking.

**Comment by manugarri** [comment]
@ballcap231 has a point, epochs is not the same thing as time, is there any way to maybe add a callback "on_epoch_end" to check the elapsed time?

**Comment by w4nderlust** [comment]
This can turn in a feature request, but it is not a common pattern in machine learning, you usually specify either the number of epochs or the number of steps.

**Comment by manugarri** [comment]
@w4nderlust definitely. But in AutoML, it is the most common way to regulate how to train.

**Comment by w4nderlust** [comment]
@manugarri can you provide some authoritative links about it? I'm considering it and want to be sure :)

**Comment by manugarri** [comment]
Sure! 
- [autosklearn `time_left_for_this_task` example from the repo](https://github.com/automl/auto-sklearn/blob/master/examples/example_random_search.py#L74)
- [H2O `max_runtime_secs` docs ](http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/algo-params/max_runtime_secs.html) 
- [TPOT `max_time_mins` in the repo](https://github.com/EpistasisLab/tpot/blob/master/tpot/base.py#L113)
- [AutoKeras `time_limit` example from the repo](https://github.com/keras-team/autokeras/blob/master/examples/a_simple_example/mnist.py#L9)

**Comment by w4nderlust** [comment]
Thank you so much. Will reopen this issue and trsnform it into a feature request.

**Comment by iXanthos** [comment]
@w4nderlust Do we have an update on this enhancement? It is vital for comparing Ludwig with other AutoML systems and as @manugarri mentioned, the common ground is `max_runtime`.

**Comment by w4nderlust** [comment]
@iXanthos Ludwig is not an AutoML system, not sure where you got that >( Anyway, yes, this is a feature on our radar, but we have a few more with higher priority. This is not actually difficult to implement anyway, if you care about it I can give you precise directions on how to implement it and do a PR about it.

**Comment by iXanthos** [comment]
> @iXanthos Ludwig is not an AutoML system, not sure where you got that from, certainly not from me :) Anyway, yes, this is a feature on our radar, but we have a few more with higher priority. This is not actually difficult to implement anyway, if you care about it I can give you precise directions on how to implement it and do a PR about it.

AutoDL**. I would be interested, yeah.

**Comment by w4nderlust** [comment]
@iXanthos it's not really about the ML/DL part, it's about the Auto part :) At the moment Ludwig has no way to specify directly an hyperparameter search, the only thing it does is providing you with reasonable default for your model configuration, there is nothing automatic other than the model composition, so I definitely wouldn't call it AutoML or AutoDL. At least not yet, as hyperopt is coming.

Regarding how to implement the timeout, it's pretty simple: you start a timer around line 427 of `models/model.py` at the beginning of the training, and then after each batch of training (line 580) you check if the timer is above the maximum time (you have to add an additional parameter for that in the train function signature), if that's true then you break, and you also want to add this check in the evaluation calls (608, 614, 623) and at the end of the epoch (741). Alternatively, you can create a scheduled timer that fires when passes the maximum time and calls a function that halts the training, similar to `set_epochs_to_1_or_quit`. Let me know if this is clear and if you need more details.

**Comment by tgaddair** [comment]
Hey @iXanthos @manugarri @ballcap231, a few updates on this thread:

1. Ludwig does now support a (very experimental) AutoML API you're welcome to try out [here](https://github.com/ludwig-ai/ludwig/tree/master/ludwig/automl). We will be releasing this officially in Ludwig 0.6, but you're welcome to try it out now or contribute if interested.
2. There is now a way to specify the training time both in the AutoML using the `time_limit_s` param, as well as when using hyperopt with the Ray Tune executor. For the latter, you can set `hyperopt.executor.time_budget_s` in the config.
3. For the programmatic API, Ludwig also supports a [Callback](https://github.com/ludwig-ai/ludwig/blob/master/ludwig/callbacks.py) interface now, which you can use to implement custom early stopping behavior, such as a custom time limit or other criteria.

With these added features, I'll go ahead and consider this issue closed, but feel free to follow-up with any additional questions.
